docs/python/programming-guide.md

```diff
@@ -7,236 +7,386 @@ pagination_prev: quickstart
+from dbos import DBOS
+from fastapi import FastAPI
+
+from schema import example_table
+
+app = FastAPI()
+DBOS(fastapi=app)
 
-```python showLineNumbers
 @DBOS.transaction()
-def insert_greeting(name: str) -> str:
-    query = dbos_hello.insert().values(name=name)
-    DBOS.sql_session.execute(query)
-    logging.info(f">>> STEP 2: Greeting to {name} recorded in the database!")
+def insert_row():
+        DBOS.sql_session.execute(example_table.insert().values(name="dbos"))
+
+@DBOS.transaction()
+def count_rows():
+    count = DBOS.sql_session.execute(example_table.select()).rowcount
+    print(f"Row count: {count}")
+
+@app.get("/")
+@DBOS.workflow()
+def dbos_workflow():
+    insert_row()
+    count_rows()
 ```
 
-[`@DBOS.transaction()`](./tutorials/transaction-tutorial.md) is a special annotation for workflow steps that access the database.
-It executes your function in a single database transaction.
-We recommend using transactions because:
+This workflow first inserts a new row into your table, then prints the total number of rows in into your table.
+The database operations are done in DBOS _transactions_. These are a special kind of step optimized for database accesses.
+They execute as a single database transaction and give you access to a pre-configured database client (`DBOS.sql_session`).
+Learn more about transactions [here](./tutorials/transaction-tutorial.md).
+
+Now, start your app with `dbos start`, then visit this URL: http://localhost:8000.
+
+You should see an output like:
 
-1. They give you access to a pre-configured database client (`DBOS.sql_session`), which is more convenient than connecting to the database yourself. You no longer need to configure a SQLAlchemy engine!
-2. Under the hood, transactions are highly optimized because DBOS can update its record of your program's execution _inside_ your transaction. For more info, see our ["how workflows work"](../explanations/how-workflows-work.md) explainer.
+```shell
+INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
+Row count: 1
+```
 
-Now, restart your app with `dbos start` and visit its URL again: http://localhost:8000/greeting/Mike.
-The app should durably execute your workflow the same as before!
+Every time you visit http://localhost:8000, your workflow should insert another row, and the printed row count should go up by one.
 
-The code for this guide is available [on GitHub](https://github.com/dbos-inc/dbos-demo-apps/tree/main/python/greeting-guestbook).
+Congratulations!  You've finished the DBOS Python guide.
+You can find the code from this guide in the [DBOS Toolbox](https://github.com/dbos-inc/dbos-demo-apps/tree/main/python/dbos-toolbox) template app.
 
-Next, to learn how to build more complex applications, check out our Python tutorials and [example apps](../examples/index.md).
+Next, to learn how to build more complex applications, check out the Python tutorials and [example apps](../examples/index.md).
```

Comment:
At the end, add a block that contains all code (in a details block). Then provide pointers to related tutorials pages.